### DISCUSSION TITLE
Wagtail multi-tenancy with django-tenants

### DISCUSSION THREAD

**Discussion by NotANormalNerd** [wagtail/wagtail#12653] [Show and tell]
# Wagtail multi-tenancy with django-tenants

## Introduction
There has recently been a new wave of discussion in regard to [multi tenancy in Wagtail](https://github.com/wagtail/wagtail/discussions/7521). While Wagtail [supports multi-site and multi-instance out of the box,](https://docs.wagtail.org/en/latest/advanced_topics/multi_site_multi_instance_multi_tenancy.html) multi-tenancy is not supported as it would require a lat of separation and changes to the codebase. 

As we are running and maintaining a Wagtail multi-tenant projects, I was asked to give a wider explanation of how we do it and which experiences we have made.

## [django-tenants](https://github.com/django-tenants/django-tenants)

Since Wagtail is based on Django, we have a whole ecosystem of possibilities available to us, including the [django-tenants](https://github.com/django-tenants/django-tenants) package. This package allows us to have multiple tenants in a single Django application and provides multi-tenancy on a database level, which restricts us to postgres.

django-tenants uses the [schemas functionality of postgres](https://www.postgresql.org/docs/current/ddl-schemas.html) to separate the data of each tenant. This means that each tenant has its own schema in the database, with a public tenant containing all the shared data.

This allows us to decide per application which data should be shared between tenants and which data should be separated, but also means that we have to make sure no shared data links to any tenant data, as this would essentially break the foreign key relationship.

Now to the why: 
* django-tenants is the simplest solution to the problem of multi-tenancy if you are already building a Django application on postgres. Moving the separation onto the database level, allows us to cleanly separate the data, without many changes to our codebase.
* I personally already built two SaaS applications with django-tenants and have plenty of experiences using it, which makes me familiar with all the ins and outs of the package.
* The package is actively maintained and has already reached a very mature state. django-tenants is feature complete and will probably stay where it is.

## How we use django-tenants

The simplest approach to using django-tenants with Wagtail, or any other Django app, is to have all data and all applications as tenant-specific and only have the actual tenant data in the public schema. This means that all data is separate per tenants, pages, images, documents, etc. are all tenant-specific.

```python
# settings.py
TENANT_APPS = [
    # Django
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.sitemaps',
    'django.contrib.sessions',
    # Our apps
    'cms.home',
    # Wagtail
    'wagtail.embeds',
    'wagtail.sites',
    'wagtail.snippets',
    'wagtail.documents',
    'wagtail.images',
    'wagtail.search',
    'wagtail.admin',
    'wagtail',
]

SHARED_APPS = [
    'django_tenants',
    'cms.tenants',
]
```
Specific instructions and explanations can be found in the [django-tenants documentation](https://django-tenants.readthedocs.io/en/latest/install.html).

But it would also be relatively easy to have Wagtail users shared between tenants and add some more specific permissions to the users to restrict access to certain tenants.

Let's say we want to implement a shared event database, we would have to define the events app as a shared app and add the necessary permissions to the users.

```python
# settings.py
TENANT_APPS = [
    # Django
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.sitemaps',
    'django.contrib.sessions',
    # Our apps
    'cms.home',
    # Wagtail
    'wagtail.embeds',
    'wagtail.sites',
    'wagtail.snippets',
    'wagtail.documents',
    'wagtail.images',
    'wagtail.search',
    'wagtail.admin',
    'wagtail',
]

SHARED_APPS = [
    'django_tenants',
    'cms.tenants',
    'cms.events',
]
```

### Running the API on the public schema
Since the API itself doesn't really have any models, there is no need to really think about the database layer. The only thing that we have to do is to map a parameter to a tenant and make sure we are in the parent context when we move on.

## Caveats
One thing that would probably have to be checked is multi-tenancy with multi-site, as Wagtail has its own multi-site functionality. This would probably require some changes to the Wagtail codebase, but I am not sure about that.

It would also (probably) not be possible to share specific page types between tenants, since they all have a foreign key relation to the page model, which would need to be shared.

There are also some performance optimizations in Wagtail, especially the use of `@cached_property` that will leak data between tenants. This is because the cache is per process and not per tenant. The same applies for all external services, like redis.

As mentioned by @cnk in the discussion mentioned above, there is also the problem that because we have different schemas, we will have to migrate all the schema separately, which is okay for 2 - 30 tenants, but will be difficult for 100 tenants.

## Conclusion
Wagtail multi-tenancy is possible with django-tenants, but it requires some changes to the codebase and some considerations about the performance and the migration process. The best idea would probably to start off with a supported multi-tenancy scenario and work from there.

In the end django-tenants is a simple solution to a quite complex problem, even though the learning curve can be a bit steep. But once you have had the "now you are thinking in tenants/schemas", it is quite easy to work with it and requires little to no changes to the codebase.

Please feel free to ask questions and be specific about your usecase, I will help with my experience as much as I can.

**Comment by thibaudcolas** [comment]
Thank you! Super valuable to see this written as I’ve hear about django-tenants but have never actually seen it used yet. If there’s a an example demo project out there that’d be great to see.

> There are also some performance optimizations in Wagtail, especially the use of @cached_property that will leak data between tenants. This is because the cache is per process and not per tenant. The same applies for all external services, like redis.

Is that something you’ve had to implement workarounds for then?

**Reply by NotANormalNerd** [reply]
Yes, since we have a internal Wagtail fork with some customizations, so we just replaced the `@cached_property` with a normal `@property`.

I will check if I can put together a `bakerydemo-tenants` in a reasonable amount of time for you to test and get the hang of it.

**Comment by HadiAssadDiab** [comment]
Hi @NotANormalNerd,

You wrote:
> There are also some performance optimizations in Wagtail, especially the use of @cached_property that will leak data between tenants. This is because the cache is per process and not per tenant. The same applies for all external services, like redis.

Is there any simple solution to avoid this from happening? Also, I am using Elasticsearch. Will using Django tenants also cause trouble with Elasticsearch?

**Reply by NotANormalNerd** [reply]
Hey, `@cached_property` is a python/django internal solution and will always use memory. So this will also affect you even when you use elasticsearch

I don't know you specific usecase for using elasticsearch (headless/not headless...) so maybe if you provide more info, I can give a better opinion.

**Reply by HadiAssadDiab** [reply]
Hey again. First off, thank you very much for the quick reply. I really appreciate it. 

Please correct me if I am wrong, but as far as I understand, Wagtail only uses caching if it is manually added by the developer. 
If I do not add caching, will my Wagtail multi-tenant instance then work without issues? 

I am using Django tenants, Wagtail, Elasticsearch, and PostgreSQL.
